experimental/list_formatter/src/list_formatter.rs

```diff
@@ -0,0 +1,223 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use formatted_string_builder::SimpleFormattedStringBuilder;
+use regex::Regex;
+use crate::patterns::get_patterns;
+
+#[derive(Debug)]
+pub enum Error {
+    UnknownLocale,
+}
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum FieldType {
+    Element,
+    Literal,
+}
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum Type {
+    And,
+    Or,
+    Unit,
+}
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum Width {
+    Wide,
+    Short,
+    Narrow,
+}
+
+pub struct ListFormatter<'a> {
+    first: Pattern<'a>,
+    pair: Pattern<'a>,
+    middle: Pattern<'a>,
+    last: Pattern<'a>,
+}
+
+impl<'a> ListFormatter<'a> {
+    pub fn new(locale: &str, type_: Type, width: Width) -> Result<ListFormatter<'static>, Error> {
+        match get_patterns(locale, type_, width) {
+            None => Err(Error::UnknownLocale),
+            Some(raw) => { 
+                // let [first, pair, middle, last] = raw.map(Pattern::parse);
+                // Ok(ListFormatter{first, pair, middle, last})
+                Ok(ListFormatter {
+                    first: Pattern::parse(raw[0]), 
+                    pair: Pattern::parse(raw[1]), 
+                    middle: Pattern::parse(raw[2]), 
+                    last: Pattern::parse(raw[3])
+                })
+            }
+        }
+    }
+
+    fn format_internal<B>(
+        &self,
+        values: &[&str],
+        empty: fn() -> B,
+        single: fn(&str) -> B,
+        apply_pattern: fn(&str, &PatternParts<'a>, B) -> B,
+    ) -> B {
+        match values.len() {
+            0 => empty(),
+            1 => single(values[0]),
+            2 => apply_pattern(values[0], &self.pair.get_parts(values[1]), single(values[1])),
+            n => {
+                let mut builder = apply_pattern(values[n-2], &self.last.get_parts(values[n-1]), single(values[n-1]));
+                for i in (1..n - 2).rev() {
+                    builder = apply_pattern(values[i], &self.middle.get_parts(values[i+1]), builder);
+                }
+                apply_pattern(values[0], &self.first.get_parts(values[1]), builder)
+            }
+        }
+    }
+
+    pub fn format(&self, values: &[&str]) -> String {
+        self.format_internal(
+            values,
+            || "".to_string(),
+            |value| value.to_string(),
+            |value, (before, between, after), mut builder| { 
+                builder = builder + after;
+                builder.insert_str(0, between);
+                builder.insert_str(0, value);
+                builder.insert_str(0, before);
+                builder
+            }
+        )
+    }
+
+    pub fn format_to_parts(&self, values: &[&str]) -> SimpleFormattedStringBuilder<FieldType> {
```

Comment:
Praise: You came up with a very simple, elegant algorithm, better than ICU4C/ICU4J.